### PR TITLE
Add a basic DDS image plugin

### LINKED ISSUE #252: Request DDSImagePlugin
I would like to request the .dds(direct draw surface) format be decoded and added as an image plugin. I guess this is the right place for requests...?

I found this as a start. It doesn't do all dxt formats but some basic ones.
http://nullege.com/codes/show/src%40p%40i%40pilgrim-HEAD%40pilgrim%40codecs%40__init__.py/9/dds.DDS/python

### PR DESCRIPTION
**Pull request by jleclanche**
Only supports DXT1 and DXT5 for now.
The pixel formats ideally should be supported in decode.c instead,
but for now this is good enough.

Fixes #252

### CODE REVIEW

**Comment by hugovk** [comment]
Thanks for the contribution!

Please could you include unit tests? (See [here](https://github.com/python-pillow/Pillow/blob/master/CONTRIBUTING.md).)

That way we can know it works on all Python versions and other operating systems, and to make sure it doesn't get broken in the future.

**Comment by wiredfool** [comment]
@jleclanche  Thanks for pushing this in as a PR.

Quick set of reliability changes:  

 * [ ] need to remove the main() and run the tests from the test suite
 * [ ] Can't have an assert in there, check and raise a SyntaxError or IOError
 * [ ] The unpack( ,fp.read(n)) should be checked for length and an IOError thrown if the length is not appropriate ((after _ensure_read in TiffImagePlugin.ImageFileDirectory_v2) (which should probably be pulled out into _binary or _util))
 * [ ] need to add some documentation at least in the image-file-formats.rst file. 

If you don't get to these, It's something that I'll get to at some point.

File: PIL/DdsImagePlugin.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+                    elif cc == 3:
+                        r, g, b = _c3(r0, r1), _c3(g0, g1), _c3(b0, b1)
+
+                    idx = 4 * ((y + j) * width + (x + i))
+                    ret[idx:idx+4] = bytes([r, g, b, alpha])
+
+    return bytes(ret)
+
+
+class DDSImageFile(ImageFile.ImageFile):
```
Comment (hugovk) [inline PIL/DdsImagePlugin.py]: Please name like DdsImageFile.

**Comment by jleclanche** [comment]
Did the easy bits. I'll probably leave it up to you for the rest @wiredfool if you don't mind... feeling a bit nauseated right now ;)

**Comment by wiredfool** [comment]
no worries

**Comment by hugovk** [comment]
@jleclanche Do you have some test images we can distribute under the PIL licence?

**Comment by wiredfool** [comment]
@hugovk -- there are the ones here: http://code.qt.io/cgit/qt/qtimageformats.git/tree/tests/shared/images/dds?id=a4670f609c48fed61cbacf1a0a780ea27483ce18

But I think they're some form of gpl, either gplv2, or 3, or lgpl 2.1 or 3, or fdl.

**Comment by hugovk** [comment]
I've sent an email to ask if we can use them under the PIL licence.

**Comment by Metallicow** [comment]
@wiredfool I got a almost complete set of test images you can use for this and for whatever for that matter. How would you like them. I have 50 saved in various dxt levels with a psd also used with the nvidia photoshop plugin to make them all.
Should I do a PullReqz where at??? or can a 7z be attached to a post like here?

@jleclanche Thanks for your work on this!

**Comment by wiredfool** [comment]
@Metallicow You attach a zip like archive here.

**Comment by Metallicow** [comment]
@wiredfool Here ya go. 
These where made a few years back when I was tinkering with writing a python plugin but I don't think the photoshop plugin has changed any. They should all be labeled correctly.
License CC0/Public domain/PIL whatever.
[ddstests.zip](https://github.com/python-pillow/Pillow/files/78870/ddstests.zip)

**Comment by hugovk** [comment]
See https://github.com/python-pillow/Pillow/pull/1650 for this PR with added tests.

**Comment by jleclanche** [comment]
Closing in favour of #1650 which contains my commit

**Pull request closed without merging**